### PR TITLE
feature #427: created attach selective discipline endpoint

### LINKED ISSUE #427: Take selective discipline
The task is to create endpoint which receives array of disciplines that user takes as their own discipline.

There are a couple of conditions that must be checked before saving these disciplines to database:
- amount of disciplines. as we can see, there are different amount of disciplines (1 for 3rd and 1/2 for 4th semester, 3/4 for 5th and 6th semester, 3 for 7th and 8th semester). You need to create new table in database, which stores information about amount of selective disciplines, allowed for specific group in each semester.
- their semester and year. you should split given array by semester and year and check each array's amount separately

Body is an array of discipline ids

### PR DESCRIPTION
**Pull request by f-ogoog**
- created endpoint which attach selective discipline to User
- added AlreadySelectedException, ExcessiveSelectiveDisciplinesException, NotApprovedException, NotBelongToGroupException exceptions
- added AttachSelectiveDisciplinesDTO
- added SelectiveDisciplinesPipe StudentByApprovedPipe
- added swagger documentation to this endpoint
- added integration tests
- added DbGroup and GroupMapper
- changed includes in GroupRepository

closes #427

### CODE REVIEW

File: src/v2/api/pipes/StudentByApprovedPipe.ts [hunk 1]
```diff
@@ -0,0 +1,21 @@
+import { Injectable, PipeTransform } from '@nestjs/common';
+import { StudentRepository } from '../../database/repositories/StudentRepository';
+import { State } from '@prisma/client';
+import { NotApprovedException } from '../../utils/exceptions/NotApprovedException';
+
+@Injectable()
+export class StudentByApprovedPipe implements PipeTransform {
```
Comment (basarabst) [inline src/v2/api/pipes/StudentByApprovedPipe.ts:7]: change naming

**Review by basarabst — CHANGES_REQUESTED** [review]

File: src/v2/api/services/DisciplineService.ts [hunk 2]
```diff
@@ -29,4 +37,108 @@ export class DisciplineService {
...
+  }
+
+  private checkAlreadySelectedDisciplines (disciplines, selectedDisciplines) {
+    const isUnique = checkIfArrayIsUnique([...disciplines, ...selectedDisciplines]);
+    if (!isUnique) {
+      throw new AlreadySelectedException();
+    }
+  }
+
+  private async checkExcessiveSelectiveDisciplines (sortedDisciplines, sortedSelectedDisciplines, groupId) {
```
Comment (just-hoshinon) [inline src/v2/api/services/DisciplineService.ts:85]: types:
```ts
type SortedDisciplines = {
  year: number;
  semester: number;
  disciplines: string[];
}
```

File: src/v2/api/services/DisciplineService.ts [hunk 3]
```diff
@@ -29,4 +37,108 @@ export class DisciplineService {
...
+    for (const period of sortedDisciplines) {
```
Comment (just-hoshinon) [inline src/v2/api/services/DisciplineService.ts:86]: can be simplified to: 
```ts
for (const { year, semester, disciplines } of pendingDisciplines) {
    const pendingAmount = disciplines.length;
    const selectedAmount = selectedDisciplines
    .find((p) => p.year === year && p.semester === semester)
    ?.disciplines.length ?? 0;
  ```

File: src/v2/api/services/DisciplineService.ts [hunk 4]
```diff
@@ -29,4 +37,108 @@ export class DisciplineService {
     });
     return this.disciplineTeacherMapper.getDisciplineTeachersWithTeacherParams(disciplineTeachers);
   }
+
+  private getSortedDisciplinesByPeriod (disciplines) {
```
Comment (just-hoshinon) [inline src/v2/api/services/DisciplineService.ts:41]: type

File: src/v2/api/services/DisciplineService.ts [hunk 5]
```diff
@@ -29,4 +37,108 @@ export class DisciplineService {
...
+        },
+      });
+      const { amount } = selectiveAmounts[0];
+      if (selectiveAmount + userSelectedAmount > amount) {
+        throw new ExcessiveSelectiveDisciplinesException();
+      }
+    }
+  }
+
+  private async attachSelectiveDisciplines (userId, disciplines) {
```
Comment (just-hoshinon) [inline src/v2/api/services/DisciplineService.ts:106]: types

File: src/v2/api/services/DisciplineService.ts [hunk 6]
```diff
@@ -29,4 +37,108 @@ export class DisciplineService {
...
+        selectiveDisciplines: {
+          createMany: {
+            data: disciplines.map((disciplineId) => ({ disciplineId })),
+          },
+        },
+      },
+    });
+  }
+
+  async selectDisciplines (userId, body) {
```
Comment (just-hoshinon) [inline src/v2/api/services/DisciplineService.ts:121]: types

**Review by just-hoshinon — CHANGES_REQUESTED** [review]

**Review by just-hoshinon — APPROVED** [review]

**Review by just-hoshinon — APPROVED** [review]

**Review by just-hoshinon — APPROVED** [review]

**Review by basarabst — APPROVED** [review]

**Pull request merged**